File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.Stack;
public class Main{
    public static void main(String[] args) throws IOException{
           BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
       
        while(true){
         String input = br.readLine();
            if(input.equals(".")) break;
            if(isBalanced(input)){
                sb.append("yes\n");
            }else{
                sb.append("no\n");
            }
        }
         System.out.println(sb);       
     }
    
    private static boolean isBalanced(String input){
        Stack<Character> stack = new Stack<>();
        for(char c : input.toCharArray()){
            if(c=='(' || c =='['){
                stack.push(c);
            }else if((c==')'||c==']')){
                  if(stack.isEmpty()) return false;
                
                char top = stack.pop();
                if((c==')' && top != '(')|| (c==']' && top != '[')){
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.c
#include <stdio.h>

char stack[101];
int top = -1;

int emptyStack()
{
	return top == -1;
}

void push(char ch)
{
	stack[++top] = ch;
}

char pop() 
{
	return stack[top--];
}



int isOk(char *str)
{
	char ch, chs;
	
	top = -1;
	
	for (int i = 0; str[i] != '\0'; i++) {
		ch = str[i];
		if (ch == '(' || ch == '[') push(ch);
		else if (ch == ')' || ch == ']') {
			if (emptyStack()) return 0;
			chs = pop();
			if (ch == ')' && chs != '(' ||
				ch == ']' && chs != '[') return 0;
		}
	}
	
	if (emptyStack()) return 1;
	return 0;
}

int main()
{
	int n;
	char str[101];
	
	gets(str);
	while (str[0] != '.') {
		if (isOk(str)) puts("yes");
		else puts("no");
		
		gets(str);
	}
	

}